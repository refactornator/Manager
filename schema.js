export default (schema = {
  schema: [
    {
      name: 'Note',
      primaryKey: 'key',
      properties: {
        key: 'string',
        text: 'string',
        createdAt: 'date',
        report: { type: 'Report', optional: true }
      }
    },
    {
      name: 'Report',
      primaryKey: 'key',
      properties: {
        key: 'string',
        name: 'string',
        notes: { type: 'list', objectType: 'Note' }
      }
    }
  ]
});
