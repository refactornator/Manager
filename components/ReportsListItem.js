import {
  AlertIOS,
  Dimensions,
  LayoutAnimation,
  TouchableOpacity
} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import React, { Component } from 'react';
import styled from 'styled-components/native';
import uuid from 'uuid';

const { width: windowWidth } = Dimensions.get('window');

export default class ReportsListItem extends Component {
  addReport = name => {
    const { realm } = this.props;

    try {
      realm.write(() => {
        const newReport = realm.create('Report', {
          key: uuid.v1(),
          name
        });
      });
      LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
    } catch (e) {
      console.log('Error on creation', e);
    }
  };

  render() {
    const { navigation, realm, data } = this.props;

    if (typeof data === 'object') {
      return (
        <TouchableOpacity onPress={() => navigation.navigate('Report', data)}>
          <Item>
            <Icon name="user" size={44} color="#366ddc" />
            <Name>{data.name}</Name>
          </Item>
        </TouchableOpacity>
      );
    } else if (data === 'add') {
      return (
        <TouchableOpacity
          onPress={() => {
            AlertIOS.prompt('Enter Name', null, this.addReport, 'plain-text');
          }}
        >
          <Item>
            <Icon name="plus-circle" size={44} color="#366ddc" />
            <Name>Add Report</Name>
          </Item>
        </TouchableOpacity>
      );
    }
  }
}

const Item = styled.View`
  flex: 1;
  width: ${windowWidth / 2}px;
  height: ${windowWidth / 2}px;
  align-items: center;
  justify-content: center;
`;

const Name = styled.Text`
  color: #366ddc;
  margin-top: 6px;
  font-size: 20px;
`;
