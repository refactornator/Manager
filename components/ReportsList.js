import { Dimensions } from 'react-native';
import Grid from 'react-native-grid-component';
import React, { Component } from 'react';
import styled from 'styled-components/native';

import ReportsListItem from './ReportsListItem';

const { width: windowWidth } = Dimensions.get('window');

export default class ReportsList extends Component {
  renderItem = (data, index) => {
    const { navigation } = this.props;
    const { realm } = this.props.screenProps;

    return <ReportsListItem key={index} {...{ navigation, realm, data }} />;
  };

  render() {
    const { realm } = this.props.screenProps;
    const reports = realm.objects('Report').sorted('name');
    const data = [...reports, 'add'];

    return (
      <Container>
        <Grid
          style={{ flex: 1 }}
          renderItem={this.renderItem}
          data={data}
          itemsPerRow={2}
        />
      </Container>
    );
  }
}

const Container = styled.View`
  flex: 1;
  background-color: #f4f4f4;
`;
