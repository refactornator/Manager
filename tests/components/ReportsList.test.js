import React from 'react';

import { shallow } from 'enzyme';

import Realm from '../mocks/realmMock';
import ReportsList from '../../components/ReportsList';
import ReportsListItem from '../../components/ReportsListItem';
import schema from '../../schema';

const realm = new Realm(schema);
const data = [{ key: 1, name: 'William' }];
realm.prepareData('Report', data);

const props = {
  navigation: { navigate: () => {} },
  screenProps: { realm }
};

it('renders a list of reports', () => {
  const render = shallow(<ReportsList {...props} />);
  expect(render).toMatchSnapshot();
});

it('has a renderItem function for each item', () => {
  const render = shallow(<ReportsList {...props} />);

  const index = 0;
  expect(render.instance().renderItem(data[index], index)).toEqual(
    <ReportsListItem
      key={index}
      navigation={props.navigation}
      realm={realm}
      data={data[index]}
    />
  );
});
