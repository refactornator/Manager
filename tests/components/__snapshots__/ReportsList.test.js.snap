// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of reports 1`] = `
<Styled(View)>
  <Grid
    data={
      Array [
        Object {
          "key": 1,
          "name": "William",
          "notes": Array [],
        },
        "add",
      ]
    }
    itemHasChanged={[Function]}
    itemsPerRow={2}
    onEndReached={[Function]}
    refreshControl={null}
    renderFooter={[Function]}
    renderItem={[Function]}
    renderPlaceholder={null}
    renderSectionHeader={[Function]}
    sections={false}
    style={
      Object {
        "flex": 1,
      }
    }
  />
</Styled(View)>
`;
